### PR TITLE
Issue 38803: Update VCS info reporting to mothership

### PR DESCRIPTION
**Pull request by labkey-jeckels**
Submit VCS info for all modules in JSON metrics, in both check-for-updates ping and exception reporting

Capture Core module's info in mothership.SoftwareReleases table

Collect branch and tag info, and handle Git revision numbers. Move BuildTime from ServerSessions to SoftwareReleases

### CODE REVIEW

File: mothership/src/org/labkey/mothership/query/MothershipSchema.java [hunk 1]
```diff
@@ -325,33 +342,18 @@ public FilteredTable createExceptionStackTraceTable(ContainerFilter cf)
...
-        ExprColumn maxRevisionLookupColumn = new ExprColumn(result, "MaxSVNRevisionLookup", getSoftwareReleaseSQL("DESC", true), JdbcType.INTEGER);
-        maxRevisionLookupColumn.setFk(softwareReleaseFK);
-        maxRevisionLookupColumn.setDescription("Highest SVN revision (and corresponding SVN branch). Slower, but full lookup to mothership.SoftwareRelease");
-        result.addColumn(maxRevisionLookupColumn);
-
-        ExprColumn minRevisionColumn = new ExprColumn(result, "MinSVNRevision", getSoftwareReleaseSQL("ASC", false), JdbcType.VARCHAR);
-        minRevisionColumn.setDescription("Lowest SVN revision (and corresponding SVN branch)");
+        ExprColumn minRevisionColumn = new ExprColumn(result, "MinRevision", getSoftwareReleaseSQL("ASC"), JdbcType.INTEGER);
+        minRevisionColumn.setFk(softwareReleaseFK);
+        minRevisionColumn.setDescription("Most oldest release with a report");
```
Comment (labkey-susanh) [inline mothership/src/org/labkey/mothership/query/MothershipSchema.java]:
```suggestion
        minRevisionColumn.setDescription("Oldest release with a report");
```

File: api/src/org/labkey/api/util/UsageReportingLevel.java [hunk 2]
```diff
@@ -277,10 +264,25 @@ protected void putModulesMetrics(Map<String, Object> runningMetrics)
...
     }
 
-    protected void putModuleControllerHits(Map<String, Object> runningMetrics)
+    public static void putModulesBuildInfo(Map<String, Map<String, Object>> allModulesStats)
+    {
+        for (Module module : ModuleLoader.getInstance().getModules())
+        {
+            Map<String, Object> moduleStats = allModulesStats.computeIfAbsent(module.getName(), k -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER));
+            Map<String, Object> moduleBuildInfo = new HashMap<>();
+            moduleBuildInfo.put("buildTime", module.getBuildTime());
```
Comment (labkey-ankurj) [inline api/src/org/labkey/api/util/UsageReportingLevel.java]: Where are these two (moduleStats and moduleBuildInfo) local maps being used?

File: mothership/src/org/labkey/mothership/MothershipManager.java [hunk 3]
```diff
@@ -160,63 +160,71 @@ public void insertException(ExceptionStackTrace stackTrace, ExceptionReport repo
...
 
-            if (svnURL == null)
-            {
-                filter.addCondition(FieldKey.fromString("SVNURL"), null, CompareType.ISBLANK);
-            }
-            else
+            if (buildNumber == null)
             {
-                filter.addCondition(FieldKey.fromString("SVNURL"), svnURL);
+                buildNumber = fabricateDescription(container, revision, url, branch, tag, buildTime);
```
[the remark this answers is not in the document]
Comment (labkey-jeckels) [reply mothership/src/org/labkey/mothership/MothershipManager.java]: This is indeed messy. I'm trying to strike a balance among:
Being semi-compatible with descriptions we've already generated and stored in the DB
Handling submissions from servers running 19.2 and older builds that don't include all of the same data
Capturing enough data to be useful for 19.3 servers

Open to suggestions for improvement

File: api/src/org/labkey/api/util/UsageReportingLevel.java (hunk 2 above)
Comment (labkey-jeckels) [reply api/src/org/labkey/api/util/UsageReportingLevel.java]: Added some comments to help clarify.

**Comment by labkey-jeckels** [comment]
@labkey-ankurj Can you approve my changes, or let me know that my comments didn't provide enough detail? Your requests are blocking the normal merge (though I can override as an admin) even though Susan approved.

**Pull request merged**